helpers/mysql/lib/opentelemetry/helpers/mysql.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0module OpenTelemetry
+
+module OpenTelemetry
+  module Helpers
+    # This class contains helpers for MySQL instrumentation libraries, like mysql2 and
+    # trilogy. It is intended for use by instrumentation developers and not to
+    # be called directly by an application.
+    #
+    # To use this in your instrumentation, the `Instrumentation` class for
+    # your gem must contain configuration options for:
+    #  * `:span_name`
+    #  Example:
+    #  `option :span_name, default: :statement_type, validate: %I[statement_type db_name db_operation_and_name]`
+    #
+    # @api public
+    class MySQL
+      QUERY_NAMES = [
+        'set names',
+        'select',
+        'insert',
+        'update',
+        'delete',
+        'begin',
+        'commit',
+        'rollback',
+        'savepoint',
+        'release savepoint',
+        'explain',
+        'drop database',
+        'drop table',
+        'create database',
+        'create table'
+      ].freeze
+
+      QUERY_NAME_REGEX = Regexp.new("^(#{QUERY_NAMES.join('|')})", Regexp::IGNORECASE)
+      EXTRACT_STATEMENT_FAILURE_MESSAGE = 'Error extracting SQL statement type'
+
+      # This is a span naming utility intended for use in MySQL database
+      # adapter instrumentation.
+      #
+      # @param sql [String] The SQL statement for the span.
+      # @param operation [String] The database operation.
+      # @param database_name [String] The name of the database.
+      # @param config [Hash] The user's configuration for the database adapter.
+      #  Desired keys:
+      #    * `:span_name` => A symbol describing the type of name desired. Expected options are `:statement_type`, `:db_name`, and `:db_operation_and_name`. A nil or unknown `:span_name` will return 'mysql' as the span name
+      # @return [String] The span name.
+      # @api public
+      def self.database_span_name(sql, operation, database_name, config)
+        case config[:span_name]
+        when :statement_type
+          extract_statement_type(sql)
+        when :db_name
+          database_name
+        when :db_operation_and_name
+          db_operation_and_name(operation, database_name)
+        end || 'mysql'
+      end
+
+      # @api private
+      def self.extract_statement_type(sql)
+        QUERY_NAME_REGEX.match(sql) { |match| match[1].downcase } unless sql.nil?
+      rescue StandardError => e
+        OpenTelemetry.handle_error(message: "#{EXTRACT_STATEMENT_FAILURE_MESSAGE}: #{e.message}", exception: e)
+        "OpenTelemetry Error: #{EXTRACT_STATEMENT_FAILURE_MESSAGE}"
```

Comment:
@arielvalentin - Calling `OpenTelemetry.handle_error` here will raise the error after it's reported. This causes the span to have a `nil` name value. This is the behavior I'm seeing when I run the tests, but I might be misinterpreting the impact of using `.handle_error`.

When we call `OpenTelemetry.logger.<level>`, the return value is `true`, which allows an error to be logged, but does not stop the span from getting the default name of `'mysql'`.

I think it's preferable to get a span with a generic name over raising an error in an application. 

Do you see any issues with how I've added `.handle_error` that might be interfering with the span naming process? If this is indeed the behavior, would you be open to reverting the change and using `OpenTelemetry.logger.error` instead?

Reply:
@arielvalentin  - After thinking through this a little more, I think the first release of these helper gems should include the original behavior. 

`handle_error` changes the behavior by raising the error and preventing span creation with the default name of `mysql`.

We can update the gem to use `handle_error` here instead in a future PR, and I'm happy to add a comment and make a ticket to address this work in the future.

 Are you comfortable with this approach?

Reply:
Sure.